### PR TITLE
raftstore: add more checks before proposing joint confchange

### LINKED ISSUE #9445: raftstore::test_joint_consensus::test_restart_in_joint_state failed
Latest build: <a href="https://internal.pingcap.net/idc-jenkins/job/tikv_ghpr_test/39932/display/redirect">https://internal.pingcap.net/idc-jenkins/job/tikv_ghpr_test/39932/display/redirect</a>

### PR DESCRIPTION
**Pull request by NingLin-P**
### What problem does this PR solve?

Issue Number: close #9445 #9332 #8789 #8775

Problem Summary:

Currently, TiKV doesn't check leave joint command. If the leader is `DemotingVoter`, it will step down after applying the leave joint command and lead to region unavailable until a new leader elect. Normally, this should not happened, because PD (with its acknowledge about the region) won't send such command to the leader,  and as such command can send to the leader with the right `RegionEpoch` meaning PD has up-to-date acknowledge. But we should add such check for robustness.

### What is changed and how it works?

What's Changed:

This PRs add checks to prevent leave joint while the leader is `DemotingVoter`, and also add checks to prevent split or merge region while the region is in joint state, again this also should not happen because of the epoch check.

This PRs also fix some unstable tests,  which cause by checking `must_get_none` on peers that were added as learner with joint confchange. This PRs fix them by checking the roles of these peers.

### Check List 

Tests 

- Integration test

### Release note 
- raftstore: add more checks before proposing joint confchange

### CODE REVIEW

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -3283,17 +3299,32 @@ where
...
-                "peer_id" => self.fsm.peer_id(),
-                "message" => ?msg,
-                "err" => %e,
-                "error_code" => %e.error_code(),
-            );
-            cb.invoke_with_response(new_error(e));
-            return;
+        if msg.has_admin_request() {
+            if let Err(e) = self.check_joint_state(&msg) {
+                warn!(
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:3304]: Does PD follows the same constraint?

File: components/raftstore/src/store/fsm/peer.rs [hunk 2]
```diff
@@ -3107,6 +3107,22 @@ where
         self.ctx.store_stat.engine_total_keys_written += metrics.written_keys;
     }
 
+    fn check_joint_state(&self, msg: &RaftCmdRequest) -> Result<()> {
+        if !self.fsm.peer.in_joint_state() {
+            return Ok(());
+        }
+        let cmd = msg.get_admin_request();
+        if cmd.has_prepare_merge() || cmd.has_commit_merge() || cmd.has_split() || cmd.has_splits()
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:3115]: Shouldn't the pipeline checker filter them automatically?

**Comment by gengliqi** [comment]
I suggest fixing https://github.com/tikv/tikv/issues/3572 in this PR.
Maybe it's better to mention this issue is about correctness, not availability.

File: components/raftstore/src/store/fsm/peer.rs (hunk 1 above)
Comment (NingLin-P) [reply components/raftstore/src/store/fsm/peer.rs:3304]: Yes, when a region in joint state PD won't try to split/merge this region /cc @HunDunDM

File: components/raftstore/src/store/fsm/peer.rs (hunk 2 above)
Comment (gengliqi) [reply components/raftstore/src/store/fsm/peer.rs:3115]: Do you mean epoch checker? The epoch checker can only filter the concurrent admin cmd.

Comment (gengliqi) [reply components/raftstore/src/store/fsm/peer.rs:3115]: Actually both the merge and split cmd are ok. I agree that rejecting the merge cmd for simpleness. But maybe it's not necessary to reject split cmd?

Comment (NingLin-P) [reply components/raftstore/src/store/fsm/peer.rs]: Yes, I removed the check for split cmd, PTAL again

File: components/raftstore/src/store/fsm/peer.rs [hunk 3]
```diff
@@ -3115,6 +3115,14 @@ where
             return Ok(());
         }
 
+        if self.fsm.peer.in_joint_state() {
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:3118]: I think you should only prevent prepare merge otherwise it can be an incompatible change.

**Review by BusyJay — APPROVED** [review]
Either fix #3572 in this PR or a separate one LGTM.

**Review by gengliqi — APPROVED** [review]

**Comment by NingLin-P** [comment]
> Either fix #3572 in this PR or a separate one LGTM.

I will fix it in the next PR

**Pull request merged**